src/combinatorics/hall/basic.lean

```diff
@@ -0,0 +1,250 @@
+/-
+Copyright (c) 2021 Alena Gusakov, Bhavik Mehta, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alena Gusakov, Bhavik Mehta, Kyle Miller
+-/
+import combinatorics.hall.finite
+import topology.category.Top.limits
+
+/-!
+# Hall's Marriage Theorem
+
+Given a list of finite subsets $X_1,X_2,\dots,X_n$ of some given set
+$S$, P. Hall in [Hall1935] gave a necessary and sufficient condition for
+there to be a list of distinct elements $x_1,x_2,\dots,x_n$ with
+$x_i\in X_i$ for each $i$: it is when for each $k$, the union of every
+$k$ of these subsets has at least $k$ elements.
+
+Rather than a list of finite subsets, one may consider indexed families
+`t : ι → finset α` of finite subsets with `ι` a `fintype`, and then the list
+of distinct representatives is given by an injective function `f : ι → α`
+such that `∀ i, f i ∈ t i`, called a *matching*.
+This version is formalized as `finset.all_card_le_bUnion_card_iff_exists_injective'`
+in a separate module.
+
+The theorem can be generalized to remove the constraint that `ι` be a `fintype`.
+As observed in [Halpern1966], one may use the constrained version of the theorem
+in a compactness argument to remove this constraint.
+The formulation of compactness we use is that inverse limits of nonempty finite sets
+are nonempty (`nonempty_sections_of_fintype_inverse_system`), which uses the
+Tychonoff theorem.
+The core of this module is constructing the inverse system: for every finite subset `ι'` of
+`ι`, we can consider the matchings on the restriction of the indexed family `t` to `ι'`.
+
+## Main statements
+
+* `finset.all_card_le_bUnion_card_iff_exists_injective` is in terms of `t : ι → finset α`.
+* `fintype.all_card_le_rel_image_card_iff_exists_injective` is in terms of a relation
+  `r : α → β → Prop` such that `rel.image r {a}` is a finite set for all `a : α`.
+* `fintype.all_card_le_filter_rel_iff_exists_injective` is in terms of a relation
+  `r : α → β → Prop` on finite types, with the Hall condition given in terms of
+  `finset.univ.filter`.
+
+## Todo
+
+* The statement of the theorem in terms of bipartite graphs is in preparation.
+
+## Tags
+
+Hall's Marriage Theorem, indexed families
+-/
+
+open finset
+
+universes u v
+
+/-- The sup directed order on finsets.
+
+TODO: remove when #9200 is merged.  There are two ways `finset α` can
+get a `small_category` instance (used in
+`hall_matchings_functor`). The first is from the preorder on `finset
+α` and the second is from this `directed_order`. These categories
```

Comment:
So wait, we can construct a small category from a directed order? Given that a `directed_order` is a `preorder`, why do we even have this construction in the first place?